### ISSUE TITLE
[FEATURE] Support storage network for RWX volumes

### ISSUE THREAD

**Issue by c3y1huang** [longhorn/longhorn#8184]
## Is your feature request related to a problem? Please describe (👍 if you like this request)

Previously, when the storage network was introduced, the RWX volume feature was experimental. Now that this feature has graduated from its experimental status, Longhorn should also support the storage network for RWX volumes.

## Describe the solution you'd like

- Implement storage networking for RWX volumes.

## Describe alternatives you've considered

`None`

## Additional context

`None`

**Comment by c3y1huang** [comment]
Last month, Mutus decided to archive the mutlus-service repository. So, we need to explore how the share-manager can operate, with or without Kubernetes service using the secondary network interface created by Multus.

- https://github.com/k8snetworkplumbingwg/multus-service-archived
- https://github.com/k8snetworkplumbingwg/multus-cni/issues/466

**Comment by c3y1huang** [comment]
> Last month, Mutus decided to archive the mutlus-service repository. So, we need to explore how the share-manager can operate, with or without Kubernetes service using the secondary network interface created by Multus.

Exploring the possibility with EndpointSlice. (WIP)

**Comment by c3y1huang** [comment]
> Exploring the possibility with [EndpointSlices](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#:~:text=In%20Kubernetes%2C%20an%20EndpointSlice%20contains,that%20match%20the%20Service%20selector.).

We can leverage custom Kubernetes Endpoint, headless Service, and service FQDN to integrate storage-network with RWX volumes.

Below is the PoC supporting this approach.
```
> k -n longhorn-system get ep
NAME                                       ENDPOINTS                                         AGE
longhorn-engine-manager                    <none>                                            28m
longhorn-replica-manager                   <none>                                            28m
longhorn-frontend                          10.42.1.4:8000,10.42.2.5:8000                     28m
longhorn-backend                           10.42.0.16:9500,10.42.1.15:9500,10.42.2.15:9500   28m
longhorn-recovery-backend                  10.42.0.16:9503,10.42.1.15:9503,10.42.2.15:9503   28m
longhorn-admission-webhook                 10.42.0.16:9502,10.42.1.15:9502,10.42.2.15:9502   28m
longhorn-conversion-webhook                10.42.0.16:9501,10.42.1.15:9501,10.42.2.15:9501   28m
pvc-93213f2d-2d76-458a-ab68-485987ea60eb   192.168.3.4:2049                                  44s

> kubectl -n longhorn-system get volume -o yaml | grep shareEndpoint
    shareEndpoint: nfs://pvc-93213f2d-2d76-458a-ab68-485987ea60eb.longhorn-system.svc.cluster.local/pvc-93213f2d-2d76-458a-ab68-485987ea60eb

> k get pod
NAME                        READY   STATUS    RESTARTS   AGE
rwx-test-79d5494fd8-wlh9r   2/2     Running   0          3m36s
rwx-test-79d5494fd8-cq94d   2/2     Running   0          3m36s
rwx-test-79d5494fd8-m8jnt   2/2     Running   0          3m36s



```

**Comment by c3y1huang** [comment]
Part of the [this PoC](https://github.com/longhorn/longhorn/issues/8184#issuecomment-2097746378) was achieved by removing HostPID from the CSI plugin DaemonSet. After talking with the team and some investigation, we can observe that setting HostPID to false resulted in the NFS mount point becoming unresponsive after restarting the CSI plugin pod.

To integrate the storage network with RWX volume, both Kubernetes and host networking must be visible to the CSI plugin.

So now here is the challenge. In the current implementation, the mounter utilized nsenter to execute the mount within the host namespace. This won't work for the storage network because the process lacks awareness of the Kubernetes network where Multus exists.

Replacing hostPID, nsenter with hostNetwork won't work either because, Multus then loses control over managing the pod network, and the pod ends up lacking Multus networking.

So far, the only working solution is restarting the share manager pod when the CSI plugin pod is restarted and triggers workload pod deletion. This allows the volume to be re-published to the workload pod. 

But, this solution isn't ideal. Need more investigation.

**Comment by c3y1huang** [comment]
After more digging into why the NFS client connection became unresponsive after the CSI plugin pod restarted. My theory is that the kernel's NFS client is still connected through the container's network namespace, leading to the broken mount since the namespace is already gone. The theory comes from the observation that `/proc/fs/nfsfs/servers` and `/proc/fs/nfsfs/volumes` are not reflected on the host when the mount is not executed in the host network namespace. These two files are also cleared after the restart.

Based on this theory, the kernel's NFS client is unaware that the client namespace is tied to the CSI plugin pod, resulting in a dangling mount. To re-establish the connection, it is necessary to trigger a re-mount (umount, then mount). Therefore, I think the best approach is to restart the workload pod whenever the CSI plugin pod restarts. However, this will cause temporary interruptions to the workload pod. 

Alternatively, we could implement an init-container in the CSI plugin pod to handle the remount of all discovered NFS mounts, but IMO, letting Kubernetes handle the remounts through the volume-attaching lifecycle is a safer approach.

**Comment by c3y1huang** [comment]
After speaking with @ejweber , we believe the best solution is to restart the workload pod. However, since this introduces a new behavior where the workload pod will restart when the CSI plugin pod restarts, we should limit this behavior to the pod using the RWX volume + storage network.

We've also discussed implementing another mechanism to handle the umount/mount process, such as using an init-container or a service in a DaemonSet. However, this adds another level of complexity, and as mentioned in my previous comment, this approach is not ideal because the mounting must include all the correct options.

**Comment by derekbit** [comment]
> this approach is not ideal because the mounting must include all the correct options.

@c3y1huang Can you elaborate more on this statement?

**Comment by c3y1huang** [comment]
> > this approach is not ideal because the mounting must include all the correct options.
> 
> @c3y1huang Can you elaborate more on this statement?

The approach of remounting is not ideal because we need to include all the correct mount options. When remounting the NFS, the main challenge is not in the umounting, but the mounting. Specifically, during remounting, we need to detect and retain all the existing NFS mount options. For example, `Flags: rw, relatime, vers=4.1, rsize=1048576, wsize=1048576, namlen=255, softerr, softreval, noresvport, proto=tcp, timeo=600, retrans=5, sec=sys, clientaddr=10.42.2.47, local_lock=none, addr=10.43.171.45`. If any of these options are omitted or incorrectly specified during the remount, it could lead to mount failure.

**Comment by derekbit** [comment]
> > > this approach is not ideal because the mounting must include all the correct options.
> > 
> > 
> > @c3y1huang Can you elaborate more on this statement?
> 
> The approach of remounting is not ideal because we need to include all the correct mount options. When remounting the NFS, the main challenge is not in the umounting, but the mounting. Specifically, during remounting, we need to detect and retain all the existing NFS mount options. For example, `Flags: rw, relatime, vers=4.1, rsize=1048576, wsize=1048576, namlen=255, softerr, softreval, noresvport, proto=tcp, timeo=600, retrans=5, sec=sys, clientaddr=10.42.2.47, local_lock=none, addr=10.43.171.45`. If any of these options are omitted or incorrectly specified during the remount, it could lead to mount failure.

I see. The options we need to provide are only `rw, softerr, softreval, noresvport, timeo=600, retrans=5`, and other options are automatically appended by the mount process.

**Comment by c3y1huang** [comment]
> I see. The options we need to provide are only `rw, softerr, softreval, noresvport, timeo=600, retrans=5`, and other options are automatically appended by the mount process.

I see. But I prefer not to hardcode the options and instead retain them from the existing NFS mount points.

**Comment by derekbit** [comment]
> > I see. The options we need to provide are only `rw, softerr, softreval, noresvport, timeo=600, retrans=5`, and other options are automatically appended by the mount process.
> 
> I see. But I prefer not to hardcode the options and instead retain them from the existing NFS mount points. And I cannot judge if the mount options would vary after an upgrade. IMO, it's safer to let the CSI controller handle this through the Kubernetes volume attachement/detachment flow.

Sure. Just explained the mount options specified by Longhorn.
`it's safer to let the CSI controller handle this through the Kubernetes volume attachement/detachment flow.` sounds good to me.

**Comment by Vicente-Cheng** [comment]
Sorry for the late update here. Generally having the same conclusion as above.

TL;DR, NFS tied to the network namespace since 2011~2012.

I do some checks as below,
1. After the CSI-plugin pod respawns, the NFS server cannot receive any packets.

2. Check the RPC logs as below:
```
May 20 10:23:36.219427 harvester-node-0 kernel: RPC:       registering /proc/net/rpc <-- first time provision
.
.
.
May 20 10:25:45.859443 harvester-node-0 kernel: RPC:       registering /proc/net/rpc <-- csi-plugin respawn
May 20 10:25:46.339552 harvester-node-0 kernel: RPC:       xs_tcp_send_request(264) = 0
May 20 10:25:47.111484 harvester-node-0 kernel: RPC:       xs_connect scheduled xprt 000000000cc04fe6
May 20 10:25:47.111877 harvester-node-0 kernel: RPC:       worker connecting xprt 000000000cc04fe6 via tcp to 10.53.48.53 (port 2049)
May 20 10:25:47.117095 harvester-node-0 kernel: RPC:       000000000cc04fe6 connect status 101 connected 0 sock state 7
May 20 10:25:47.117352 harvester-node-0 kernel: RPC:       xs_close xprt 000000000cc04fe6
May 20 10:25:47.123688 harvester-node-0 kernel: RPC:       xs_tcp_state_change client 000000000cc04fe6...
May 20 10:25:47.123842 harvester-node-0 kernel: RPC:       state 7 conn 0 dead 0 zapped 1 sk_shutdown 3
May 20 10:25:50.183466 harvester-node-0 kernel: RPC:       xs_connect scheduled xprt 000000000cc04fe6
May 20 10:25:50.183630 harvester-node-0 kernel: RPC:       worker connecting xprt 000000000cc04fe6 via tcp to 10.53.48.53 (port 2049)
May 20 10:25:50.183661 harvester-node-0 kernel: RPC:       000000000cc04fe6 connect status 101 connected 0 sock state 7
May 20 10:25:50.187730 harvester-node-0 kernel: RPC:       xs_close xprt 000000000cc04fe6
May 20 10:25:50.187822 harvester-node-0 kernel: RPC:       xs_tcp_state_change client 000000000cc04fe6...
May 20 10:25:50.187843 harvester-node-0 kernel: RPC:       state 7 conn 0 dead 0 zapped 1 sk_shutdown 3
```
Seems the second csi-plugin pod would like to re-register the RPC, but it failed.
That might cause the NFS client to use the old NS instead of a new one. (I am not sure of that; just thought that might be related)

Some links might be related to the namespace introduction.
https://lore.kernel.org/netdev/20120113084545.31750.83655.stgit@localhost6.localdomain6/
https://lore.kernel.org/netdev/20120110121246.7307.1435.stgit@localhost6.localdomain6/

So, the whole flow might look as follows:
1. csi-plugin pod provisioned.
2. create a nfs client A (with specific ns)
3. csi-plugin somehow dead (in our case, we killed the pod) then respawn
4. create a new nfs client B (with another ns)

Somehow the RPC client register failed, so the nfs client could not use the latest namespace.

So, we would like to re-establish the connection through a remount. It should be OK for general cases, but I am still worried about the case with a heavy workload. The remount operation might bring potential risk because something needs to reset (like the client cache, etc.).

But `it's safer to let the CSI controller handle this through the Kubernetes volume attachment/detachment flow.`
It's a good point. Thanks!

Or could we have any chance to keep the hostPID and provide the service entry point through the multus network?

**Comment by c3y1huang** [comment]
> Or could we have any chance to keep the hostPID and provide the service entry point through the multus network?

Once the process enters the host network namespace, it won't be able to see the Multus network and Kubernetes services unless the Kubernetes DNS is added to /etc/resolv.conf (this has not been tested yet). Therefore, I believe our best option for now is to restart the workload pod without fiddling host DNS resolver.

I am curious about the risks associated with restarting the workload pod. If problems arise from remounting during the restart, these issues should likely be from day 1, or will eventually surface and need to be addressed. We need to resolve these issues because I think it's not uncommon for the Kubelet or users to restart the workload pod.

**Comment by chriscchien** [comment]
Verified **pass** on longhorn master(longhorn-manager `81d46a`) 

1. Install longhorn v1.6.2
2. Setup storage network ([steps](https://longhorn.github.io/longhorn-tests/manual/release-specific/v1.6.0/test-storage-network/))
3. Create a RWX volume, write data to pod mount point
4. Did not see `kube-system/demo-192-168-0-0` in all the share-manager pod `metadata.annotations`
5. Did not see `kube-system/demo-192-168-0-0` in all the longhorn-csi-plugin pod `metadata.annotations`
6. Upgrade to v1.7.0-dev
7. Set `storage-network-for-rwx-volume-enabled` to `true`
8. Create another RWX volume, write data to pod mount point
9. Observed `kube-system/demo-192-168-0-0` in all the share-manager pods `metadata.annotations`
10. Observed `kube-system/demo-192-168-0-0` in all the longhorn-csi-plugin pods `metadata.annotations`
```
metadata:
  annotations:
    k8s.v1.cni.cncf.io/network-status: |-
      [{
          "name": "cbr0",
          "interface": "eth0",
          "ips": [
              "10.42.2.51"
          ],
          "mac": "9e:bd:22:1c:5b:09",
          "default": true,
          "dns": {},
          "gateway": [
              "10.42.2.1"
          ]
      },{
          "name": "kube-system/demo-192-168-0-0",
          "interface": "lhnet1",
          "ips": [
              "192.168.3.7"
          ],
          "mac": "0a:ff:c3:57:dd:19",
          "dns": {}
      }]
    k8s.v1.cni.cncf.io/networks: '[{"namespace": "kube-system", "name": "demo-192-168-0-0",
      "interface": "lhnet1"}]'
```
11. All RWX workloads R/W working
12. Set `storage-network-for-rwx-volume-enabled` to `false`
11. Annotation `kube-system/demo-192-168-0-0` in observed steps 9, 10 are gone.
12. All RWX workloads R/W working.
13. Repeat step7 to step 12 on fresh installed v1.7.0-dev worked well too.

**Comment by c3y1huang** [comment]
@chriscchien could you help to review the test cases for this feature? Thank you.
- https://github.com/longhorn/longhorn-tests/pull/1931
- https://github.com/longhorn/longhorn-tests/pull/1921
